containers/src/Data/Tree.hs

```diff
@@ -207,33 +211,82 @@ instance Foldable Tree where
       where go !z (Node x ts) = foldl' go (f z x) ts
     {-# INLINE foldl' #-}
 
-    foldr1 f = go id
-      where go k (Node x ts) = foldr (\n k' prev -> f prev (go k' n)) k ts x
-    {-# INLINE foldr1 #-}
+    foldr1 = foldrMap1 id
 
-    foldl1 f (Node x ts) = foldl (foldl f) x ts
+    foldl1 = foldlMap1 id
 
     null _ = False
     {-# INLINE null #-}
 
     elem = any . (==)
     {-# INLINABLE elem #-}
 
-    maximum = foldl1' max
+    maximum = foldlMap1' id max
     {-# INLINABLE maximum #-}
 
-    minimum = foldl1' min
+    minimum = foldlMap1' id min
     {-# INLINABLE minimum #-}
 
-    sum = foldl1' (+)
+    sum = foldlMap1' id (+)
     {-# INLINABLE sum #-}
 
-    product = foldl1' (*)
+    product = foldlMap1' id (*)
     {-# INLINABLE product #-}
 
-foldl1' :: (a -> a -> a) -> Tree a -> a
-foldl1' f = \(Node x ts) -> foldl' (foldl' f) x ts
-{-# INLINE foldl1' #-}
+#if MIN_VERSION_base(4,18,0)
+-- | Folds in preorder
```

Comment:
Add a `@since FIXME` annotations (as seems to be elsewhere too).

Reply:
Thanks, missed that.